### PR TITLE
bootstrap: Introduce --no-download flag

### PR DESCRIPTION
**Pull request by bgamari**
When working on [ghcs.nix](https://gitlab.haskell.org/bgamari/ghcs-nix) I found it helpful to be able to explicitly request that the bootstrap script not automatically download tarballs to ensure that I seeded them correctly myself.

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users have been recorded in the changelog (add file to `changelog.d` directory).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: bootstrap/bootstrap.py [hunk 1]
```diff
@@ -334,8 +353,13 @@ def main() -> None:
                         help='bootstrap dependency file')
     parser.add_argument('-w', '--with-compiler', type=Path,
                         help='path to GHC')
+    parser.add_argument('--no-network', action='store_true',
+                        help='disable automatic downloading of source files')
     args = parser.parse_args()
 
+    global network_allowed
```
Comment (phadej) [inline bootstrap/bootstrap.py:360]: Please, if we manage to thread every other argument, we can do without `global`.

Comment (bgamari) [reply bootstrap/bootstrap.py]: Eliminated `global`.

File: bootstrap/bootstrap.py [hunk 2]
```diff
@@ -118,33 +131,39 @@ def verify_sha256(expected_hash: SHA256Hash, f: Path):
...
         assert cabal_sha256 is not None
+        cabal_file = TARBALLS / f'{package}.cabal'
         url = package_cabal_url(package, version, revision)
-        with urllib.request.urlopen(url) as resp:
-            shutil.copyfileobj(resp, cabal_file.open('wb'))
-            verify_sha256(cabal_sha256, cabal_file)
+        fetch_file(cabal_file, url, sha256 = cabal_sha256,
+                   network_allowed = network_allowed)
 
     return (tarball, cabal_file)
```
Comment (phadej) [inline bootstrap/bootstrap.py:168]: this returns `None` now?

I guess it still works, as cabal file is not copied if revision is `None`.

However, I don't like that we fallback here and return something. Let us be explicit and have

```python
if revision is not None:
    ...
    return (tarball, cabal_file)
    
else:
    return (tarball, None)
```

and change

```python
        # Update cabal file with revision
        if dep.revision is not None:
            shutil.copyfile(cabal_file, sdist_dir / f'{dep.package}.cabal')
```

to have

```python
assert cabal_file is not None
```

**Comment by arrowd** [comment]
As a FreeBSD packager, I'm also interested in this.

**Review by gbaz — APPROVED** [review]

**Comment by arrowd** [comment]
What it preventing this from getting merged? How can I help?

File: bootstrap/bootstrap.py [hunk 3]
```diff
@@ -118,33 +131,39 @@ def verify_sha256(expected_hash: SHA256Hash, f: Path):
...
+    print(f'Fetching {package}-{version}...')
     tarball = TARBALLS / f'{package}-{version}.tar.gz'
-    if not tarball.exists():
-        print(f'Fetching {package}-{version}...')
-        tarball.parent.mkdir(parents=True, exist_ok=True)
-        url = package_url(package, version)
-        with urllib.request.urlopen(url) as resp:
-            shutil.copyfileobj(resp, tarball.open('wb'))
-
-    verify_sha256(src_sha256, tarball)
```
Comment (phadej) [inline bootstrap/bootstrap.py:138]: the `verify_sha256` calls seem to be gone. That is not good, they are sanity check.

**Review by jneira — APPROVED** [review]
lgtm, thanks

**Comment by jneira** [comment]
> What it preventing this from getting merged? How can I help?

it needed a second approval

**Review by Kleidukos — APPROVED** [review]
LGTM :rocket:

**Comment by Kleidukos** [comment]
@bgamari Just gotta resolve the conflict and I think we should be good.

**Comment by Kleidukos** [comment]
Marking this PR as draft :slightly_smiling_face:

**Comment by bgamari** [comment]
@Kleidukos pointed out that it might be useful to rebase this. However, looking at the current state of the bootstrap script, I don't think it's strictly necessary. The user can already opt out of downloading by providing the `--bootstrap-sources` flag.

**Comment by angerman** [comment]
Is this documented somewhere?

**Comment by ulysses4ever** [comment]
@angerman https://github.com/haskell/cabal/blob/master/bootstrap/README.md has a paragraph starting with `For offline builds` and mentioning `--bootstrap-sources`. Is that enough in your view?

**Comment by ulysses4ever** [comment]
I also submitted https://github.com/haskell/cabal/pull/9096 to improve that README's formatting a bit.

**Comment by ulysses4ever** [comment]
I also submitted 

- https://github.com/haskell/cabal/pull/9096 

with improved markup, so that offline builds stand out more.

Optimistically closing...

**Pull request closed without merging**